### PR TITLE
Mirror Quantum Volume Circuits

### LINKED ISSUE #1788: Add mirror quantum volume circuits to benchmarks
## Issue Description
Using quantum volume circuits as mirror circuits ([QV][QV^dagger]) can be a simple way to provide benchmarks.   

## Proposed Solution
Add a module in `mitiq.benchmarks` with a function called
```
mitiq.benchmarks.mirror_quantum_volume_circuits.generate_mirror_quantum_volume_circuit()
```
The content of the function could explicitly call the [`generate_quantum_volume_circuit()`](https://mitiq.readthedocs.io/en/stable/apidoc.html#mitiq.benchmarks.quantum_volume_circuits.generate_quantum_volume_circuit) or copy some of that code. 

## Additional References
- Existing related benchmarks for [`generate_quantum_volume_circuit`](https://mitiq.readthedocs.io/en/stable/_modules/mitiq/benchmarks/quantum_volume_circuits.html) and [`generate_mirror_circuit()`](https://mitiq.readthedocs.io/en/stable/_modules/mitiq/benchmarks/mirror_circuits.html)
- Used and benchmarked [here](https://arxiv.org/abs/2303.02108)
- Covered also [here](https://docs.google.com/presentation/d/1QYxCFBS9EhA8A9ESBGAlN0CKxeisAPSR73_6Jrd1kA4/edit#slide=id.g22d303b4706_0_616) (a [Quantum Wednesday](https://github.com/unitaryfund/quantum-wednesday) talk)

### PR DESCRIPTION
**Pull request by purva-thakre**
Fixes #1788 

List of tests to add:

- [x] DIfference between odd and even depth : both should have the same output of all zero bitstrings
- [x] Check decompose option
- [x] Add  function in API docs: https://mitiq.readthedocs.io/en/stable/contributing_docs.html#automatically-add-information-from-the-api-docs
- [x] Conversion options
- [x] Check output as expected with seed and no seed

### CODE REVIEW

File: mitiq/benchmarks/mirror_qv_circuits.py [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+    return_type: Optional[str] = None,
+) -> Tuple[QPROGRAM, Sequence[Bitstring]]:
+    """Generate a mirror quantum volume circuit with the given number of qubits
+    and depth.
+
+    The generated circuit consists of a quantum volume circuit upto `depth/2`
+    layers followed by an inverse of the quantum volume portion upto `depth/2`
+    when `depth` is an even number.
+
+    When `depth` is odd, the layers will be chnaged to `depth+1`.
```
Comment (purva-thakre) [inline mitiq/benchmarks/mirror_qv_circuits.py:36]: Currently, the function works for when the circuit depth is `2* depth`. 

These lines are based on the referenced paper. The bitstring is not as expected when the depth is an odd number. Check the problem.

Comment (purva-thakre) [reply mitiq/benchmarks/mirror_qv_circuits.py:36]: Might be fixed by https://github.com/unitaryfund/mitiq/pull/1838/commits/121c4143a25dc1c756e8e2005481d7d992be6137

**Comment by purva-thakre** [comment]
@nathanshammah This is ready for review. I pretty much re-used some of the code & tests for quantum volume.

I am not sure why codecov is failing. The stuff I added seems to be covered. There are 3 unrelated files that seem to show up in the report.

**Comment by purva-thakre** [comment]
I think I figured out the issue with `codecov` based on what I read [here](https://docs.codecov.com/docs/unexpected-coverage-changes#reasons-for-unexpected-changes). 

The files it flagged are the ones that had codecov failing on them previously. I think it is related to [this previous PR](https://app.codecov.io/gh/unitaryfund/mitiq/pull/1814?src=pr&el=h1&utm_medium=referral&utm_source=github&utm_content=checks&utm_campaign=pr+comments&utm_term=unitaryfund).

**Comment by purva-thakre** [comment]
@nathanshammah this is ready for a review. The test that is failing in unrelated to this PR

File: mitiq/benchmarks/mirror_qv_circuits.py [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+    num_qubits: int,
+    depth: int,
+    decompose: bool = False,
+    seed: Optional[int] = None,
+    return_type: Optional[str] = None,
+) -> Tuple[QPROGRAM, Sequence[Bitstring]]:
+    """Generate a mirror quantum volume circuit with the given number of qubits
+    and depth as defined in :cite:`Amico_2023_arxiv`.
+
+    The generated circuit consists of a quantum volume circuit upto `depth/2`
```
Comment (Misty-W) [inline mitiq/benchmarks/mirror_qv_circuits.py:33]:
```suggestion
    The generated circuit consists of a quantum volume circuit up to `depth/2`
```

File: mitiq/benchmarks/mirror_qv_circuits.py [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+    layers followed by an inverse of the quantum volume portion upto `depth/2`
```
Comment (Misty-W) [inline mitiq/benchmarks/mirror_qv_circuits.py:34]:
```suggestion
    layers followed by an inverse of the quantum volume portion up to `depth/2`
```

File: mitiq/benchmarks/mirror_qv_circuits.py [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+    when `depth` is an even number.
+
+    When `depth` is odd, the layers will be chnaged to `depth+1`.
```
Comment (Misty-W) [inline mitiq/benchmarks/mirror_qv_circuits.py:37]:
```suggestion
    When `depth` is odd, the layers will be changed to `depth+1`.
```

File: mitiq/benchmarks/mirror_qv_circuits.py [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+
+
+    The output bit-string is always supposed to be a string of zeroes.
```
Comment (Misty-W) [inline mitiq/benchmarks/mirror_qv_circuits.py:40]:
```suggestion
    The ideal output bit-string is a string of zeroes.
```

File: mitiq/benchmarks/mirror_qv_circuits.py [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+    for i in range(len(qv_half_ops))[::-1]:
+        op_inverse = cirq.inverse(qv_half_ops[i])
+        mirror_half_circ.append(op_inverse, strategy=cirq.InsertStrategy.NEW)
+
+    circ = qv_half_circ + mirror_half_circ
+
+    # get the bitstring
+    circ_with_measurements = circ + cirq.measure(circ.all_qubits())
+    simulate_result = cirq.Simulator().run(circ_with_measurements)
+    bitstring = list(simulate_result.measurements.values())[0][0].tolist()
```
Comment (Misty-W) [inline mitiq/benchmarks/mirror_qv_circuits.py:83]: Why not simply a list of zeros?
```suggestion
    bitstring = [0] * num_qubits
```

**Review by Misty-W — APPROVED** [review]
Nice work @purva-thakre! 

Apart from a minor suggestion and fixing a couple typos, this looks ready to go.

File: mitiq/benchmarks/mirror_qv_circuits.py (hunk 6 above)
Comment (purva-thakre) [reply mitiq/benchmarks/mirror_qv_circuits.py:83]: I think I was trying to make sure I was getting the result I needed. 

I do use a list of zeros for[ testing the function.](https://github.com/unitaryfund/mitiq/pull/1838/files#diff-5d2d8d1fc5428a29239daa3d408b6c3a7c184b609b3d5c38bc66e508a2fae05fR28)

Comment (andreamari) [reply mitiq/benchmarks/mirror_qv_circuits.py:83]: Is it useful to return the expected bitstring, given that it must always be just a list of zeros?

For example, for `generate_rb_circuits()`, in which the expected bitstring is always 00..0 we only return the circuit.

Comment (purva-thakre) [reply mitiq/benchmarks/mirror_qv_circuits.py:83]: Not really. I was trying to stick to what the `quantum volume` function returns. Plus, I could check I was getting what I expected. 

I'll make the requested changes and remove the bitstring lines completely.

**Comment by Misty-W** [comment]
@nathanshammah did you want to take a look before we merge?

File: mitiq/benchmarks/mirror_qv_circuits.py [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+            unitary matrix gates into simpler gates.
+        seed: Seed for generating random circuit.
+        return_type: String which specifies the type of the returned
+            circuits. See the keys of ``mitiq.SUPPORTED_PROGRAM_TYPES``
+            for options. If ``None``, the returned circuits have type
+            ``cirq.Circuit``.
+
+    Returns:
+        A quantum volume circuit acting on ``num_qubits`` qubits.
+        A list of the bitstrings for the returned circuit.
```
Comment (andreamari) [inline mitiq/benchmarks/mirror_qv_circuits.py:54]:
```suggestion
```

**Review by andreamari — APPROVED** [review]
LGTM! Thanks @purva-thakre !

File: mitiq/benchmarks/mirror_qv_circuits.py (hunk 7 above)
Comment (purva-thakre) [reply mitiq/benchmarks/mirror_qv_circuits.py]: Thanks for catching this!

**Pull request merged**